Everywhere: Remove even more cyclic #include-s, add pre-commit hook to prevent more cycles

BenWiederhake opened the pull request:
This continues the removal of cyclic headers, and adds a pre-commit hook to prevent any future cycles. This script does cause and increase of the pre-commit running time, by about 110ms. See #15281 , in which I decrease the pre-commit running time by about 900ms, so clearly this shouldn't be a problem.

In case you're wondering why I care so much about clean header files: The cleaner they get, the easier it becomes to do some analysis and reduce the overall compilation times. This isn't Feng-Shui programming, but rather a small direct reduction of compilation times in theory, and preparation for future PRs that improve compilation times further. Also, cyclic headers are a real beast, and resolving them leads to headaches. In fact, they already did: https://github.com/SerenityOS/serenity/pull/15236

Also, see https://github.com/SerenityOS/serenity/pull/15235 where a bunch of more cycles got removed.

trflynn89 commented on AK/Array.h:
```diff
@@ -109,6 +109,27 @@ constexpr static auto iota_array(T const offset = {})
+
+template<typename T>
+template<size_t size>
+ALWAYS_INLINE constexpr Span<T>::Span(Array<T, size>& array)
```
Does anything actually need these constructors?

`Array` has an implicit `operator Span` much like `Vector` does, and that should be sufficient. I removed these constructors locally and everything builds fine. I'd rather that than awkwardly having a `Span::Span` definition in `Array.h`.

BenWiederhake replied:
Huh, I can't find any, either. I'll remove them then, good catch! :D

BenWiederhake commented:
Changes since last push:
- Changed the AK::Span/Array commit to instead remove the `Span::Span(Array)` constructor. Thanks @trflynn89!
- Reworded that commit accordingly

bgianfo commented:
Do you have any numbers on how long the new precommit check takes?

I’ll try it myself once I get to a real machine.

bgianfo commented:
> Do you have any numbers on how long the new precommit check takes?
> 
> I’ll try it myself once I get to a real machine.

Seems pretty speedy on my machine:

```
~/src/serenity-os/serenity$ hyperfine ./Meta/check-headers-acyclic.py
Benchmark 1: ./Meta/check-headers-acyclic.py
  Time (mean ± σ):      91.7 ms ±   6.8 ms    [User: 74.9 ms, System: 16.9 ms]
  Range (min … max):    80.6 ms … 103.1 ms    35 runs
```

BenWiederhake commented:
> Do you have any numbers on how long the new precommit check takes?
> 
> I’ll try it myself once I get to a real machine.

Yes, about 110ms, that's why I stated it in the first comment ^^
I see a standard deviation of about 4ms.

I timed it, and if you look into the function `get_dependencies` you'll see that I already collected all the low-hanging fruit of optimization. You can manually set `SHOW_TIMING` to True near the top of the file, then you'll see something like this:

```
$ /usr/bin/time ./Meta/check-headers-acyclic.py 
// Took 2.837721 + 82.813563 + 4.977624 = 90.628908  milliseconds to compute everything
// The second number splits into reading and parsing phases for each of the 2486 headers.
// No cycles detected in 2486 files (9414 successfully-resolved includes)
0.08user 0.02system 0:00.10elapsed 98%CPU (0avgtext+0avgdata 13024maxresident)k
0inputs+0outputs (0major+2656minor)pagefaults 0swaps
```

That's 2.8ms to run and parse `git ls-files *.h`, 82.8ms for the function `get_dependencies` (hence why I spent most of my time here), and 4.9…ms to prove that there are no cycles. The difference between reports (i.e. 0.10s - 90.6ms = around 10ms) is basically just bootup time of python, checking all the `build.ninja`s to determine the "current" build directory, pretty-printing, teardown, and rounding error.

BenWiederhake commented:
On an unrelated note, I had a lot of fun understanding and implementing Tarjan's strongly connected components algorithm :D

BenWiederhake commented:
Changes since last push:
- Rebased to resolve conflict with myself :D (file `Meta/lint-ci.sh`)

linusg commented on Meta/check-headers-acyclic.py:
```diff
@@ -0,0 +1,308 @@
+#!/usr/bin/env python3
```
Please reformat this with https://github.com/psf/black. We don't enforce this yet, but I intend to eventually as the number of python scripts in Meta/ is growing.

linusg commented on Meta/check-headers-acyclic.py:
```diff
@@ -0,0 +1,308 @@
+    assert result.stderr == ''
+    output = result.stdout.split('\n')
+    assert output[-1] == ''  # Trailing newline
+    assert len(output) > 500, 'There should be well over a thousand headers, not only {}?!'.format(len(output))
```
Use f-strings

linusg commented on Meta/check-headers-acyclic.py:
```diff
@@ -0,0 +1,308 @@
+    return output[:-1]
+
+
+def find_system_include(arch, include_filename, magic=[]):
```
https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

linusg replied:
Also, add some type annotations please :)

BenWiederhake replied:
Mutable default arguments are exactly what is being (ab)used here, so this is not a bug.

In retrospect: I agree, this deserves a comment at least, or even better: I'll convert it into a global variable, because that's what it is.

linusg commented on Meta/check-headers-acyclic.py:
```diff
@@ -0,0 +1,308 @@
+
+    # Collapse '..' into the path of `filename`.
+    # Note that this is currently not used, and doesn't work in some edgecases.
+    while relative_path.startswith('../') and filename:
+        relative_path = relative_path[len('../'):]
+        filename = os.path.dirname(filename)
```
Use https://docs.python.org/3/library/pathlib.html

BenWiederhake replied:
Why? os.path is also a standard library, and does exactly what's needed. In fact, as the comment explains: That case currently doesn't happen, so I'll replace the code by an `assert '../' not in filename`

linusg commented on Meta/check-headers-acyclic.py:
```diff
@@ -0,0 +1,308 @@
+    if not filename:
+        return relative_path
+
+    return os.path.join(filename, relative_path)
```
No `os.path` in modern python pls

linusg commented on Meta/check-headers-acyclic.py:
```diff
@@ -0,0 +1,308 @@
+
+
+if __name__ == '__main__':
+    os.chdir(os.path.dirname(__file__) + "/..")
```
again, pathlib

linusg reviewed (changes requested):
Cleanup looks great, some optinions on the python script :)
